dora/core/common/src/main/java/alluxio/membership/WorkerClusterView.java

```diff
@@ -0,0 +1,131 @@
+import alluxio.wire.WorkerIdentity;
+import alluxio.wire.WorkerInfo;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.Iterators;
+
+import java.time.Instant;
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Snapshot for a cluster view.
+ */
+@Immutable
+public final class WorkerClusterView implements Iterable<WorkerInfo> {
+  private final Instant mInstantCreated;
+  private final IndexedSet<WorkerInfo> mWorkers;
+
+  private static final IndexDefinition<WorkerInfo, WorkerIdentity> INDEX_WORKER_ID =
+      IndexDefinition.ofUnique(WorkerInfo::getIdentity);
+
+  /**
+   * Creates a cluster view with the give workers.
+   *
+   * @param workers workers in this view
+   */
+  public WorkerClusterView(Iterable<WorkerInfo> workers) {
+    mWorkers = new IndexedSet<>(INDEX_WORKER_ID);
+    for (WorkerInfo workerInfo : workers) {
+      mWorkers.add(workerInfo);
+    }
+    // Note Instant.now() uses the system clock and is NOT monotonic
+    // which is fine because we want to invalidate stale snapshots based on wall clock time
+    mInstantCreated = Instant.now();
+  }
+
+  /**
+   * Finds a worker by its ID.
+   *
+   * @param workerIdentity the ID of the worker to find
+   * @return the worker info of the given worker, or none if the worker is not found
+   */
+  public Optional<WorkerInfo> getWorkerById(WorkerIdentity workerIdentity) {
+    return Optional.ofNullable(
+        mWorkers.getFirstByField(INDEX_WORKER_ID, workerIdentity));
+  }
+
+  @Override
+  public Iterator<WorkerInfo> iterator() {
+    return Iterators.unmodifiableIterator(mWorkers.iterator());
+  }
+
+  /**
+   * Converts to a stream of {@link WorkerInfo}.
+   *
+   * @return stream of workers
+   */
+  public Stream<WorkerInfo> stream() {
+    return mWorkers.stream();
+  }
+
+  /**
+   * @return number of workers contained in the cluster view
+   */
+  public int size() {
+    return mWorkers.size();
+  }
+
+  /**
+   * @return if the cluster contains no worker
+   */
+  public boolean isEmpty() {
+    return mWorkers.isEmpty();
+  }
+
+  /**
+   * @return the time when this snapshot was created.
+   */
+  public Instant getSnapshotTime() {
+    return mInstantCreated;
+  }
+
+  /**
+   * Note that the equals implementation considers the creation timestamp to be part of the
+   * snapshot's identity. To compare two snapshots only by the contained workers,
+   * use {@link com.google.common.collect.Iterables#elementsEqual(Iterable, Iterable)}.
+   */
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    WorkerClusterView that = (WorkerClusterView) o;
+    return Objects.equals(mInstantCreated, that.mInstantCreated)
+        && Objects.equals(mWorkers, that.mWorkers);
```

Comment:
This relies on `equals()` on `Instant` and `IndexedSet`. I checked the implementations of both and i feel they are fine. But the code here doesn't look very straightforward. I guess you need some inline comments and UT to guarantee `equals()` works as expected.

Reply:
I will add UT to cover equals. What do you mean the code isn't straightforward? This is code generated by IDE.

Reply:
> What do you mean the code isn't straightforward?

I need to check how `equals()` work by looking at Instant and Map to know how equals() behaves here.

Anyways your test will ensure all is good.